graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsDataLoaderProvider.kt

```diff
@@ -87,50 +85,53 @@ class DgsDataLoaderProvider(
+                        val annotation = field.getAnnotation(DgsDataLoader::class.java)
+                        ReflectionUtils.makeAccessible(field)
 
-                    if (annotation.name == DgsDataLoader.GENERATE_DATA_LOADER_NAME) {
-                        throw DgsUnnamedDataLoaderOnFieldException(field)
-                    }
+                        if (annotation.name == DgsDataLoader.GENERATE_DATA_LOADER_NAME) {
+                            throw DgsUnnamedDataLoaderOnFieldException(field)
+                        }
 
-                    fun <T : Any> createHolder(t: T): LoaderHolder<T> = LoaderHolder(t, annotation, annotation.name)
-                    when (val get = field.get(dgsComponent)) {
-                        is BatchLoader<*, *> -> batchLoaders.add(createHolder(get))
-                        is BatchLoaderWithContext<*, *> -> batchLoadersWithContext.add(createHolder(get))
-                        is MappedBatchLoader<*, *> -> mappedBatchLoaders.add(createHolder(get))
-                        is MappedBatchLoaderWithContext<*, *> -> mappedBatchLoadersWithContext.add(createHolder(get))
-                        else -> throw InvalidDataLoaderTypeException(dgsComponent::class.java)
+                        fun <T : Any> createHolder(t: T): LoaderHolder<T> = LoaderHolder(t, annotation, annotation.name)
+                        when (val get = field.get(dgsComponent)) {
+                            is BatchLoader<*, *> -> batchLoaders.add(createHolder(get))
+                            is BatchLoaderWithContext<*, *> -> batchLoadersWithContext.add(createHolder(get))
+                            is MappedBatchLoader<*, *> -> mappedBatchLoaders.add(createHolder(get))
+                            is MappedBatchLoaderWithContext<*, *> -> mappedBatchLoadersWithContext.add(createHolder(get))
+                            else -> throw InvalidDataLoaderTypeException(dgsComponent::class.java)
+                        }
                     }
-                }
-        }
+            }
     }
 
     private fun addDataLoaderComponents() {
         val dataLoaders = applicationContext.getBeansWithAnnotation(DgsDataLoader::class.java)
-        dataLoaders.values.forEach { dgsComponent ->
-            val javaClass = AopUtils.getTargetClass(dgsComponent)
-            val annotation = javaClass.getAnnotation(DgsDataLoader::class.java)
-            val predicateField = javaClass.declaredFields.asSequence().find { it.isAnnotationPresent(DgsDispatchPredicate::class.java) }
-            if (predicateField != null) {
-                ReflectionUtils.makeAccessible(predicateField)
-                val dispatchPredicate = predicateField.get(dgsComponent)
-                if (dispatchPredicate is DispatchPredicate) {
-                    addDataLoaders(dgsComponent, javaClass, annotation, dispatchPredicate)
+        dataLoaders.values.distinctBy { AopUtils.getTargetClass(it) } // filter duplicate beans for the same class (tests)
```

Comment:
This was needed because I was seeing multiple beans (generic and root) for the same class in the context of tests.